src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java

```diff
@@ -332,4 +333,26 @@ private StudentProfilePage getProfilePageForStudent(String studentId) {
         return loginAdminToPage(profileUrl, StudentProfilePage.class);
     }
 
+    @AfterClass
+    public void classTearDown() {
+        // The data associated with the test accounts have to be removed when the test ends because
+        // the test accounts are shared across tests. In particular, the courses of test student1 account is not the same as
+        // `StudentHomePageUiTest` so it has to be cleared before running that test.
+
+        // Test data is never cleared after test in order to save datastore costs as we already remove and restore the data
+        // bundle when the test runs. This raises the question of why we do not remove the data bundle when the class
+        // ends then we do not have to remove it from startup. The reason is that in the case a test fails to tear down
+        // (fail to remove data bundle), another test should have no reason to fail.
+
+        // This means that removing the data bundle on startup is not always sufficient because a test only knows how
+        // to remove its associated data. This is the reason why `StudentHomePageUiTest` would fail if we don't remove the
+        // data bundle in this test.
+        // Extending removeDataBundle to remove data outside its associated data would introduce unnecessary complications
+        // like extra cost and now knowing exactly how much data to remove.
+
+        // Hence, since the account of user being logged in is shared in this test and `StudentHomepageUiTest` we need to
+        // explicitly remove the data bundle in both tests to avoid instability of tests
+        BackDoor.removeDataBundle(testData);
```

Comment:
Yes, removing the databundle at the end of a test is an unnecessary cost. Another reason not to remove associated data after a test is that in case of test failures, it helps to have the associated data in the datastore to debug the failure. If a test case X fails due to data created by another test Y, this is usually a problem of how test case X is designed; it is the responsibility of X to create a clean context for it to succeed. `StudentHomePageUiTest` should use  test user that is not used in any other test so that other tests cannot affect it.

Reply:
There are only 2 actual student accounts we can use so I don't think there's much we can do. Are you suggesting that we can use masqueraded accounts here?

Reply:
From what I remember, this test should not use masqueraded accounts as this is one of the few tests that verifies an actual login by a student.
If we do something here due to that constraint, then the explanation should mention the actual constraint. If there is no other option, I suppose even a direct coupling to another test is OK (e.g., delete data created by another test) if the reason is explained clearly(?). 
BTW, I haven't followed this PR closely; my comments may not be applicable to this context :-p

Reply:
@tshradheya I suppose we should document that we want to test an actual login here too.